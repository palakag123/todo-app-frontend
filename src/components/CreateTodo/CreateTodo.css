.create-todo-container{
    width: 65%;
    background-image: linear-gradient(white, rgb(23, 200, 246));
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}
.new-todo-form{
    display: flex;
    flex-direction: column;
    gap: 10px;
    width: 100%;
    align-items: center;
    padding-top: 40px;
    margin-top: auto;
}
#new-todo-input{
    width: 75%;
    min-height: 150px;
    padding: 15px;
    outline: none;
    border: none;
    border-radius: 7px;
    font-size: 20px;
    color:rgb(23, 200, 246);
    font-weight: 600;
    background-color: white;
    font-family: Arial, Helvetica, sans-serif;
}
#submit-button{
    background-color: white;
    border: none;
    outline: none;
    border-radius: 7px;
    padding: 10px;
    padding-right: 20px;
    padding-left: 20px;
    font-size: 20px;
    color: rgb(87,204,237);
    font-weight: 700;
}
#character-count{
    margin: 0px;
    color: white;
    font-size: 18px;
}
.count-char-container{
    display: flex;
    width: 75%;
    justify-content: flex-end;
}
.back-container{
    display: flex;
    padding-top: 40px;
    width: 75%;
    justify-content: flex-end;
    margin-top: auto;
    align-self: flex-end;
    padding: 10px;
}
#back-arrow{
    color: white;
}

@media only screen and (max-width: 770px){
    .create-todo-container{
        width: 100vw;
        min-height: 50vh;
    }
}
@media only screen and (max-width: 400px){
    #new-todo-input{
        min-height: 120px;
        width: 50%;
    }
}