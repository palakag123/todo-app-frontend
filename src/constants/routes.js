export const HOME_ROUTE = "/";
export const CREATE_TODO_ROUTE = "/create-todo";
