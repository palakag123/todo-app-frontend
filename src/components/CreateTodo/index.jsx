/* eslint-disable react/forbid-prop-types */
import React, { useState } from "react";
import PropTypes from "prop-types";
import "./CreateTodo.css";
import { useNavigate } from "react-router-dom";
import ArrowBackIcon from "@mui/icons-material/ArrowBack";
import { HOME_ROUTE } from "../../constants/routes";

function CreateTodo({ todos, setTodos }) {
  const [newTodo, setNewTodo] = useState("");

  const navigate = useNavigate();

  const onAddNewTodo = (event) => {
    setNewTodo(event.target.value);
  };

  const onNewTodoSubmit = (event) => {
    event.preventDefault();
    if (newTodo.length > 0) {
      setTodos([{ id: todos.length + 1, todo: newTodo }, ...todos]);
      navigate(HOME_ROUTE);
    }
  };

  const onHomeClick = (event) => {
    event.preventDefault();
    navigate(HOME_ROUTE);
  };

  return (
    <div className="create-todo-container">
      <form
        className="new-todo-form"
        onSubmit={onNewTodoSubmit}
        data-testid="newTodoForm"
      >
        <textarea
          id="new-todo-input"
          data-testid="newTodoInput"
          rows={5}
          cols={5}
          onChange={onAddNewTodo}
          value={newTodo}
          // style={newTodo.length > 100 ? { color: "red" } : {}}
          maxLength="100"
        />
        {/* <div
          contentEditable
          id="new-todo-input"
          rows={5}
          cols={5}
          onChange={onAddNewTodo}
        /> */}
        <div className="count-char-container">
          <p id="character-count" data-testid="charCount">
            {`${100 - newTodo.length} characters left`}
          </p>
        </div>
        <button type="submit" id="submit-button" data-testid="submitButton">
          SUBMIT
        </button>
      </form>
      <div className="back-container" data-testid="backArrow">
        <ArrowBackIcon
          id="back-arrow"
          data-testid="backArrowButton"
          onClick={onHomeClick}
        />
      </div>
    </div>
  );
}

CreateTodo.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      todo: PropTypes.string.isRequired,
      // eslint-disable-next-line comma-dangle
    })
  ).isRequired,
  setTodos: PropTypes.func.isRequired,
};

export default CreateTodo;
