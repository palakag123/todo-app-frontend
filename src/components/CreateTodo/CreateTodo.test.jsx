/* eslint-disable comma-dangle */
import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import CreateTodo from "./index";
import { HOME_ROUTE } from "../../constants/routes";

const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({ useNavigate: () => mockNavigate }));
const mockSetTodos = jest.fn();

describe("CreateTodo", () => {
  const mockTodos = [{ id: 1, todo: "first todo list item" }];

  beforeEach(() => {
    render(<CreateTodo todos={mockTodos} setTodos={mockSetTodos} />);
    mockSetTodos.mockClear();
    mockNavigate.mockClear();
  });

  it("should take a snapshot for CreateTodo", () => {
    const { asFragment } = render(
      <CreateTodo todos={mockTodos} setTodos={mockSetTodos} />
    );
    expect(
      asFragment(<CreateTodo todos={mockTodos} setTodos={mockSetTodos} />)
    ).toMatchSnapshot();
  });
  it("should have the required elements when CreateTodo is rendered", () => {
    expect(screen.getByTestId("newTodoInput")).toBeTruthy();
    expect(screen.getByTestId("charCount")).toBeTruthy();
    expect(screen.getByTestId("submitButton")).toBeTruthy();
    expect(screen.getByText("SUBMIT")).toBeTruthy();
    expect(screen.getByTestId("backArrow")).toBeTruthy();
  });
  it("should call mockNavigate function with the correct route value when the home button is clicked", () => {
    fireEvent.click(screen.getByTestId("backArrowButton"));
    expect(mockNavigate).toHaveBeenCalledWith(HOME_ROUTE);
  });
  it("should change the input value when input text changes", () => {
    const mockInput = "new todo";
    fireEvent.change(screen.getByTestId("newTodoInput"), {
      target: { value: mockInput },
    });
    expect(screen.getByTestId("newTodoInput").value).toBe(mockInput);
  });
  it("should set todos and navigate to home page on submit", () => {
    const mockInput = "new todo";
    fireEvent.change(screen.getByTestId("newTodoInput"), {
      target: { value: mockInput },
    });
    fireEvent.click(screen.getByTestId("submitButton"));
    expect(mockNavigate).toHaveBeenCalledWith(HOME_ROUTE);
    expect(mockSetTodos).toHaveBeenCalledTimes(1);
  });
  it("should not navigate to home page or set todos if input empty on submit", () => {
    const mockInput = "";
    fireEvent.change(screen.getByTestId("newTodoInput"), {
      target: { value: mockInput },
    });
    fireEvent.click(screen.getByTestId("submitButton"));
    expect(mockNavigate).toHaveBeenCalledTimes(0);
    expect(mockSetTodos).toHaveBeenCalledTimes(0);
  });
  it("should show chars left", () => {
    const mockInput = "ab";
    fireEvent.change(screen.getByTestId("newTodoInput"), {
      target: { value: mockInput },
    });
    // expect(screen.getByTestId("charCount")).toEqual(
    //   `${100 - mockInput.length} characters left`
    // );
    expect(
      screen.getByText(`${100 - mockInput.length} characters left`)
    ).toBeTruthy();
  });
});
