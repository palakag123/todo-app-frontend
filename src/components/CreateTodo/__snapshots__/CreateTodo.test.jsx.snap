// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTodo should take a snapshot for CreateTodo 1`] = `
<DocumentFragment>
  <div
    class="create-todo-container"
  >
    <form
      class="new-todo-form"
      data-testid="newTodoForm"
    >
      <textarea
        cols="5"
        data-testid="newTodoInput"
        id="new-todo-input"
        maxlength="100"
        rows="5"
      />
      <div
        class="count-char-container"
      >
        <p
          data-testid="charCount"
          id="character-count"
        >
          100 characters left
        </p>
      </div>
      <button
        data-testid="submitButton"
        id="submit-button"
        type="submit"
      >
        SUBMIT
      </button>
    </form>
    <div
      class="back-container"
      data-testid="backArrow"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="backArrowButton"
        focusable="false"
        id="back-arrow"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
